File: frontend/src/types/plm.ts
// ─── PLM Connection ───

export interface PLMConnection {
  connection_id: string;
  name: string;
  connector_type: string;
  config: Record<string, unknown>;
  is_active: boolean;
  product_line_code: string;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface PLMConnectionCreate {
  name: string;
  connector_type: string;
  config: Record<string, unknown>;
  product_line_code: string;
}

export interface PLMConnectionUpdate {
  name?: string;
  connector_type?: string;
  config?: Record<string, unknown>;
  is_active?: boolean;
  product_line_code?: string;
}

export interface PLMConnectionListResponse {
  items: PLMConnection[];
  total: number;
  page: number;
  page_size: number;
}

export interface PLMConnectionTestResponse {
  status: "ok" | "error";
  parts_count?: number;
  error?: string;
  error_class?: string;
}

// ─── PLM Part ───

export interface PLMPartSCLink {
  link_id: string;
  characteristic_type: string;
  status: string;
  sc_id: string | null;
  confirmed_at: string | null;
}

export interface PLMPart {
  part_id: string;
  connection_id: string;
  external_id: string;
  part_number: string;
  name: string;
  revision: string;
  material: string | null;
  specification: string | null;
  status: string;
  is_safety_related: boolean;
  is_key_characteristic: boolean;
  source_updated_at: string | null;
  product_line_code: string | null;
  plm_raw_data: Record<string, unknown> | null;
  sc_links: PLMPartSCLink[];
}

export interface PLMPartConfirmSCRequest {
  fmea_id: string;
  node_id: string;
  characteristic_type: "safety" | "key_characteristic";
}

export interface PLMPartConfirmSCResponse {
  status: string;
  sc_id: string;
  link_id: string;
}

export interface PLMPartListResponse {
  items: PLMPart[];
  total: number;
  page: number;
  page_size: number;
}

// ─── PLM BOM ───

export interface PLMBOM {
  bom_id: string;
  connection_id: string;
  external_id: string;
  parent_part_number: string;
  parent_revision: string;
  child_part_number: string;
  child_revision: string;
  quantity: number;
  bom_revision: string;
  level: number;
  source_updated_at: string | null;
  product_line_code: string | null;
  plm_raw_data: Record<string, unknown> | null;
}

export interface PLMBOMListResponse {
  items: PLMBOM[];
  total: number;
  page: number;
  page_size: number;
}

export interface PLMBOMTreeNode {
  parent_part_number: string;
  parent_revision: string;
  child_part_number: string;
  child_revision: string;
  quantity: number;
  level: number;
  bom_revision: string;
}

export interface PLMBOMTreeResponse {
  root: string;
  revision: string;
  bom_revision: string;
  items: PLMBOMTreeNode[];
  total: number;
}

export interface PLMBOMImportResponse {
  imported_nodes: number;
  imported_edges: number;
  root: string;
  revision: string;
  bom_revision: string;
  fmea_id: string;
}

// ─── PLM Change Order ───

export interface PLMChangeOrder {
  change_id: string;
  connection_id: string;
  external_id: string;
  change_number: string;
  title: string;
  description: string | null;
  change_type: string;
  status: string;
  priority: string;
  affected_part_numbers: string[];
  proposed_changes: Record<string, unknown> | null;
  requested_by: string | null;
  approved_by: string | null;
  planned_implementation_date: string | null;
  actual_implementation_date: string | null;
  source_updated_at: string | null;
  product_line_code: string | null;
  plm_raw_data: Record<string, unknown> | null;
}

export interface PLMChangeOrderListResponse {
  items: PLMChangeOrder[];
  total: number;
  page: number;
  page_size: number;
}

// ─── PLM Change Impact Task ───

export interface PLMChangeImpactTask {
  task_id: string;
  change_id: string;
  status: string;
  retry_count: number;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
  error_message: string | null;
  result: Record<string, unknown> | null;
}

// ─── PLM Dashboard ───

export interface PLMDashboard {
  part_count: number;
  bom_count: number;
  pending_ecn_count: number;
  pending_sc_count: number;
  recent_changes: PLMChangeOrder[];
}

File: backend/app/schemas/plm.py
"""PLM Pydantic v2 schemas."""

import uuid
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


CONNECTOR_TYPE_PATTERN = r"^(mock|rest|siemens_tc|dassault_enovia|ptc_windchill)$"


class PLMConnectionCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    connector_type: str = Field(..., pattern=CONNECTOR_TYPE_PATTERN)
    config: dict[str, Any] = Field(default_factory=dict)
    product_line_code: str = Field(..., min_length=1)


class PLMConnectionUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    connector_type: Optional[str] = Field(None, pattern=CONNECTOR_TYPE_PATTERN)
    config: Optional[dict[str, Any]] = None
    is_active: Optional[bool] = None
    product_line_code: Optional[str] = Field(None, min_length=1)


class PLMConnectionResponse(BaseModel):
    connection_id: uuid.UUID
    name: str
    connector_type: str
    config: dict[str, Any]
    is_active: bool
    product_line_code: str
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PLMConnectionListResponse(BaseModel):
    items: list[PLMConnectionResponse]
    total: int
    page: int
    page_size: int

    model_config = ConfigDict(from_attributes=True)


class PLMPartResponse(BaseModel):
    part_id: uuid.UUID
    connection_id: uuid.UUID
    external_id: str
    part_number: str
    name: str
    revision: str
    material: Optional[str] = None
    specification: Optional[str] = None
    status: str
    is_safety_related: bool
    is_key_characteristic: bool
    source_updated_at: Optional[datetime] = None
    product_line_code: Optional[str] = None
    plm_raw_data: Optional[dict[str, Any]] = None

    model_config = ConfigDict(from_attributes=True)


class PLMBOMResponse(BaseModel):
    bom_id: uuid.UUID
    connection_id: uuid.UUID
    external_id: str
    parent_part_number: str
    parent_revision: str
    child_part_number: str
    child_revision: str
    quantity: float
    bom_revision: str
    level: int
    source_updated_at: Optional[datetime] = None
    product_line_code: Optional[str] = None
    plm_raw_data: Optional[dict[str, Any]] = None

    model_config = ConfigDict(from_attributes=True)


class PLMChangeOrderResponse(BaseModel):
    change_id: uuid.UUID
    connection_id: uuid.UUID
    external_id: str
    change_number: str
    title: str
    description: Optional[str] = None
    change_type: str
    status: str
    priority: str
    affected_part_numbers: list[str]
    proposed_changes: Optional[dict[str, Any]] = None
    requested_by: Optional[str] = None
    approved_by: Optional[str] = None
    planned_implementation_date: Optional[datetime] = None
    actual_implementation_date: Optional[datetime] = None
    source_updated_at: Optional[datetime] = None
    product_line_code: Optional[str] = None
    plm_raw_data: Optional[dict[str, Any]] = None

    model_config = ConfigDict(from_attributes=True)


class PLMChangeImpactTaskResponse(BaseModel):
    task_id: uuid.UUID
    change_id: uuid.UUID
    status: str
    retry_count: int
    created_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None
    result: Optional[dict[str, Any]] = None

    model_config = ConfigDict(from_attributes=True)


class PLMDashboardResponse(BaseModel):
    part_count: int
    bom_count: int
    pending_ecn_count: int
    pending_sc_count: int
    recent_changes: list[PLMChangeOrderResponse]

    model_config = ConfigDict(from_attributes=True)


class BOMImportRequest(BaseModel):
    fmea_id: uuid.UUID
    overwrite: bool = False


class PLMPartLinkFMEARequest(BaseModel):
    fmea_id: uuid.UUID
    node_id: str


class PLMPartConfirmSCRequest(BaseModel):
    fmea_id: uuid.UUID
    node_id: str
    characteristic_type: Literal["safety", "key_characteristic"]
